### PR TITLE
New variable expansion syntax

### PR DESCRIPTION
**Pull request by cmouse**
- [x] Describe new syntax
- [x] Add in-depth explanation of the syntax, and how to extend it
- [x] Fix all old variables

JIRA: DOV-6924

### CODE REVIEW

File: docs/core/settings/variables.md [hunk 1]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+Basic syntax is `%{variable (| filter | filter ...)}`, which means that most existing
+variables work, but there are some changes, so check them carefully.
+
+The simple case of just getting a value of variable is `%{variable}`.
+These can be in middle of strings.
+
+Another way to get variables is `%{provider:variable}`, where the value is provided by
+a provider. There are global providers, and context-specific providers.
+
+Variable can be then filtered with various filters, such as `%{variable | upper}` to get
```
Comment (vmo69) [inline docs/core/settings/variables.md:47]: "A variable can be then filtered"

File: docs/core/config/optimization.md [hunk 2]
```diff
@@ -71,7 +71,7 @@ the most useful optimization you can do.
 - [[setting,mail_prefetch_count]] setting may be helpful with some mailbox
   formats.
 
-- [[setting,mail_volatile_path,/tmp/dovecot-volatile/%2.256Nu/%u]] moves, e.g.,
+- [[setting,mail_volatile_path,/tmp/dovecot-volatile/%{user|md5|fold(256)|hex(4)}u/%{user}]] moves, e.g.,
```
Comment (sirainen) [inline docs/core/config/optimization.md:74]: hex(2)

File: docs/core/config/guides/nfs.md [hunk 3]
```diff
@@ -64,13 +64,13 @@ Potential optimizations to use:
...
-* Use [[setting,mail_volatile_path,/dev/shm/dovecot/%2.256Nu/%u]] to store some
+* Use [[setting,mail_volatile_path,/dev/shm/dovecot/%{user | md5 | fold(256) | hex(4) }/%{user}]] to store some
   temporary files (e.g. lock files) in tmpfs rather than NFS.
-* Use [[setting,mailbox_list_index_prefix,/fast/%2.256Nu/%u]] to use "smaller
+* Use [[setting,mailbox_list_index_prefix,/fast/%{user | md5 | fold(256) | hex(4)}/%{user}]] to use "smaller
   fast storage" for index files and "larger slow storage" for mail files. Also
   use [[setting,mailbox_list_iter_from_index_dir,yes]] to list mailboxes via
   the fast index storage rather than the slow mail storage.
-* Use [[setting,mailbox_list_iter_from_index_dir,/slow/%2.256Nu/%u]] to use
+* Use [[setting,mailbox_list_iter_from_index_dir,/slow/%{user | md5 | fold(256) | hex(4)}/%{user}]] to use
```
Comment (sirainen) [inline docs/core/config/guides/nfs.md:73]: hex(2) 3 times above

File: docs/core/config/auth/master_users.md [hunk 4]
```diff
@@ -244,7 +244,7 @@ authentication. The query for that is as follows:
 passdb sql {
   query = SELECT user_name, domain_name, password \
     FROM users \
-    WHERE user_name = '%n' AND domain_name = '%d'
+    WHERE user_name = '%{username}' AND domain_name = '%{domain}'
```
Comment (sirainen) [inline docs/core/config/auth/master_users.md:247]: No more %{username} or %{domain}

File: docs/core/config/mailbox/mail_location.md [hunk 5]
```diff
@@ -27,30 +27,24 @@ See [[variable]] for a full list, but the most commonly used ones are:
...
-You can use three different kinds of hashes in [[variable]].
+Examples on how to do it:
 
-* `%N` is MD5-based "new hash" which works similarly to `%H` except it
-  gives more uniform results.
+* How to replace '%N' in new format:
 
-  * Example: `%2.256N` would return maximum 256 different hashes in range
+  * %2.256N becomes `%{ user | md5 | fold(256) | hex(2)}` would return maximum 256 different hashes in range
     `00..ff`.
```
Comment (sirainen) [inline docs/core/config/mailbox/mail_location.md:41]: Have you tested that this results in exactly the same output? If it doesn't, we break existing mail location paths. Also the old code allowed eg. `%30Nu` to get values 0..29. People might actually be using these. Would it still work with new code?

Also previously `%256Nu` gave values like 0, 1, 2, 3, .. values while `%2.256Nu` gave 00, 01, 02, 03, ... values. Can this work?

The fold() seems to be pretty much designed to replace this %N modifier. Maybe it could be changed to support all these directly without the need for |hex?

File: docs/core/config/mailbox/mail_location.md [hunk 6]
```diff
@@ -27,30 +27,24 @@ See [[variable]] for a full list, but the most commonly used ones are:
...
+| `%{user | domain}` | Domain part in `user@domain`; empty if there's no domain. |
 
 ### Directory Hashing
 
-You can use three different kinds of hashes in [[variable]].
+Examples on how to do it:
 
-* `%N` is MD5-based "new hash" which works similarly to `%H` except it
-  gives more uniform results.
+* How to replace '%N' in new format:
```
Comment (sirainen) [inline docs/core/config/mailbox/mail_location.md:38]: This is only talking about converting old format. Maybe it belongs to upgrading page? This section here should be talking about how to recommend doing the directory hashing in general. Maybe it could also mention the old ways, but it shouldn't simply start saying "How to replace old way" when people reading it wouldn't have used the old way before.

File: docs/core/settings/variables.md [hunk 7]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
 * [[link,auth_ldap]], [[link,auth_sql]], and [[link,userdb]] query strings
 * Log prefix for imap/pop3 process
 
-## Global Variables
-
-Global variables that work everywhere are:
+## Variable expansion syntax
+
+In Dovecot 2.4/3.1 we have introduced a totally new variable expansion syntax.
```
Comment (sirainen) [inline docs/core/settings/variables.md:35]: Shouldn't this be using `[[changed,var_expand_syntax]]` or something, rather than specifying explicit version numbers?

Comment (sirainen) [reply docs/core/settings/variables.md:35]: Also I think s/a totally/an entirely/ is a bit better.

File: docs/core/settings/variables.md [hunk 8]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+These can be in middle of strings.
+
+Another syntax is `%{provider:variable}`, where the value is provided by
+a provider. There are global providers, and context-specific providers.
+
+A variable can be then filtered with various filters, such as `%{variable | upper}` to get
+uppercase representation of variable. You can chain as many filters as you need.
+
+Filters can accept parameters, both positional and named. E.g. `%{literal('\r\n\')}` will expand
+to CR LF. `%{user | substr(1)}` will take first character of username.
```
Comment (sirainen) [inline docs/core/settings/variables.md:51]: Example of a named parameter would have been good.

File: docs/core/settings/variables.md [hunk 9]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+in place of variables, e.g. lookup, literal and if.
+
+When value is missing or empty, you can use the `default` filter to provide value. Missing variables
+will cause errors and must be negated with default. This does not apply to all providers, some
+providers return empty when value is missing.
+
+The new syntax also supports simple maths, you can do one operation. E.g. `%{port + 1000}`.
+Addition, substraction, multiplication, division and modulo operations are supported for now.
+
+All strings must be encapsulated with " or ', and you can escape them using \ within string.
```
Comment (sirainen) [inline docs/core/settings/variables.md:66]: with \`"\` or \`'\` quotes .. using \`\\\` within

File: docs/core/settings/variables.md [hunk 10]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+Numbers when used as parameters must be provided without quotes.
+
+## List of filters
+
+All parameters are strings unless stated otherwise.
+If parameter is any, it accepts both numbers and strings.
+Boolean type is 0 = false or 1 = true.
```
Comment (sirainen) [inline docs/core/settings/variables.md:73]: \`0\` .. \`1\`

File: docs/core/settings/variables.md [hunk 11]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+
+| Filter | Description |
+| ------ | ----------- |
+| lookup(name) | Lookup var from table. If var is variable, the name is taken from variable's contents. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:77]: var? .. or Lookup \`name\` as is in the parameter? Or `lookup(var)`?

File: docs/core/settings/variables.md [hunk 12]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+| literal(string) | Expands into literally the value. If variable is used, works like lookup. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:78]: like the \`lookup\` filter.

File: docs/core/settings/variables.md (hunk 11 above)
Comment (sirainen) [reply docs/core/settings/variables.md:77]: And what happens to input?

File: docs/core/settings/variables.md (hunk 12 above)
Comment (sirainen) [reply docs/core/settings/variables.md:78]: And what happens to input?

File: docs/core/settings/variables.md [hunk 13]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+| lower | Lowercases input. |
+| hash(method, rounds=number, salt=string) | Returns raw hash from input using given hash method. Rounds and salt are optional. |
+| md5(rounds=number, salt=string) | Alias for hash with method md5. |
+| sha1(rounds=number, salt=string) | Alias for hash with method sha1. |
+| sha256(rounds=number, salt=string) | Alias for hash with method sha256. |
+| sha384(rounds=number, salt=string) | Alias for hash with method sha384. |
+| sha512(rounds=number, salt=string) | Alias for hash with method sha512. |
+| base64(pad=boolean, url=boolean) | Base64 encode given input, defaults to pad and not url scheme. |
+| unbase64(pad=boolean, url=boolean) | Base64 decode given input, defaults to pad and not url scheme. |
+| hex(number) | Hex encode input, with optional width. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:90]: If width is larger than number, does it add 0 padding as prefix?

File: docs/core/settings/variables.md [hunk 14]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+| unhex | Hex decode input. |
+| default(value) | Replace empty or missing input with value. Clears missing variable error. If no value is provided, "" is used. |
+| reverse | Reverse input bytes. |
+| truncate(len, bits=number) | Truncate to len bytes, or number of bits. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:94]: That's confusing. How do `len` and `bits` interact? Should these have been two separate filters?

File: docs/core/settings/variables.md [hunk 15]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+| substr(start, end) | Substring bytes of input from start to end. Start and end can be negative numbers, in which case they are counted from right. End is optional. |
+| ldap_dn | Converts domain.com to dc=domain,dc=com. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:96]: \` escaping would be good all over the place in this table when referring to parameters or example values.

File: docs/core/settings/variables.md [hunk 16]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+| if(left,operator,right,true,false) | Evaluates given comparison and returns true or false value. See [conditionals](#conditionals). |
+| if(operator,right,true,false | Evaluates given comparison against input value and retuns true or false value. |
+| regexp(expression, replacement) | Performs regular expression replacement using PCRE2 syntax. Supports up to 9 capture groups. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:99]: It's not PCRE2. I think it' "extended POSIX syntax".

File: docs/core/settings/variables.md [hunk 17]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+| unbase64(pad=boolean, url=boolean) | Base64 decode given input, defaults to pad and not url scheme. |
+| hex(number) | Hex encode input, with optional width. |
+| unhex | Hex decode input. |
+| default(value) | Replace empty or missing input with value. Clears missing variable error. If no value is provided, "" is used. |
+| reverse | Reverse input bytes. |
+| truncate(len, bits=number) | Truncate to len bytes, or number of bits. |
+| substr(start, end) | Substring bytes of input from start to end. Start and end can be negative numbers, in which case they are counted from right. End is optional. |
+| ldap_dn | Converts domain.com to dc=domain,dc=com. |
+| if(left,operator,right,true,false) | Evaluates given comparison and returns true or false value. See [conditionals](#conditionals). |
+| if(operator,right,true,false | Evaluates given comparison against input value and retuns true or false value. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:98]: missing ) before |

File: docs/core/settings/variables.md [hunk 18]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+| regexp(expression, replacement) | Performs regular expression replacement using PCRE2 syntax. Supports up to 9 capture groups. |
+| number | Converts input number into input bytes. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:100]: The input must be a valid number. Also.. I think this should be either little-endian or big-endian. Now I see it is cpu-endian in the code, which is not good.

File: docs/core/settings/variables.md [hunk 19]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+| index(separator, nth) | Returns nth element from separator separated string. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:101]: starting from 0? 1? And negatives are supported too.

File: docs/core/settings/variables.md [hunk 20]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
+| username | Provides local part of user@domain value. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:102]: I think "user part". That's also how it's described elsewhere.

File: docs/core/settings/variables.md [hunk 21]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
-If [[plugin,var-expand-crypt]] is loaded, these also work globally:
-
-| Long Name | Description |
-| --------- | ----------- |
-| `encrypt; <parameters>:<field>` | Encrypt field |
-| `decrypt; <parameters>:<field>` | Decrypt field |
+| `dovecot:<name>` | Get a distribution variable, see [below](#distribution-variables) for a list of supported names. |
+| `event:<name>`   | Get an event field. Returns "" if no such field is found from event. |
+| `date:<name>`    | Get a date field, available keys are year, month, day. |
+| `time:<name>`    | Get a time field, available keys are hour, (min)ute, (sec)ond, usec. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:128]: \` quoting for all keys

File: docs/core/settings/variables.md [hunk 22]
```diff
@@ -27,24 +30,102 @@ You can use special variables in several places:
...
-| `dovecot:<name>` | Get a distribution variable, see [below](#distribution-variables) for a list of supported names.
-
-If [[plugin,var-expand-crypt]] is loaded, these also work globally:
-
-| Long Name | Description |
-| --------- | ----------- |
-| `encrypt; <parameters>:<field>` | Encrypt field |
-| `decrypt; <parameters>:<field>` | Decrypt field |
+| `dovecot:<name>` | Get a distribution variable, see [below](#distribution-variables) for a list of supported names. |
+| `event:<name>`   | Get an event field. Returns "" if no such field is found from event. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:126]: `empty string` seems clearer than `""`. Same elsewhere too.

File: docs/core/settings/variables.md [hunk 23]
```diff
@@ -118,15 +199,13 @@ See also:
...
-| | `session` | session ID for this client connection (unique for 9 years) |
-| | `auth_user` | SASL authentication ID (e.g. if master user login is done, this contains the master username). If username changes during authentication, this value contains the original username. Otherwise the same as `%{user}`. |
-| | `auth_username` | user part in `%{auth_user}` |
-| | `auth_domain` | domain part in `%{auth_user}` |
+| Variable | Description |
+| -------- | ----------- |
+| `%{user}` | Full username (e.g. user@domain) |
+| `%{user | username}` | User part in user@domain, same as `%{user}` if there's no domain |
+| `%{user | domain}` | Domain part in user@domain, empty if user with no domain |
+| `session` | Session ID for this client connection (unique for 9 years) |
```
Comment (sirainen) [inline docs/core/settings/variables.md:207]: These %{..} are now used inconsistently between tables. Some have them, some don't. Should stay consistent. In this table it's inconsistent even within the table.

File: docs/core/design/var_expand.md [hunk 24]
```diff
@@ -0,0 +1,61 @@
+---
+layout: doc
+title: Variable expansion design
+dovecotlinks:
+  var_expand: variable expansion design
+---
+
+# Variable expansion design
+
+Dovecot comes with powerful variable expansion system, which allows constructing reuseable text templates.
```
Comment (sirainen) [inline docs/core/design/var_expand.md:10]: reusable

File: docs/core/design/var_expand.md [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+
+## Design
+
+The system uses programs to perform the actual expansion. All expressions are first parsed into a program that can then be executed one or more times with parameters.
+
+Program parameters consists from variable table(s), provider(s) and escape function. A program can be executed with different parameters.
+
+## Parameters
+
+Parameters are provided via `struct var_expand_params`. Key-value mappings are provided via `const struct var_expand_table` array, which is `VAR_EXPAND_TABLE_END` terminated list of
```
Comment (sirainen) [inline docs/core/design/var_expand.md:43]: const seems a bit irrelevant here

File: docs/core/design/var_expand.md [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+key-value mappings. Key's value can also be provided by a function.
+
+Providers are used to handle scoped variables, such as passdb, ldap etc. There are also global providers which are always available.
+
+Providers can be provide with `const struct var_expand_provider` array which contains prefix and provider function, and is `VAR_EXPAND_TABLE_END` terminated.
```
Comment (sirainen) [inline docs/core/design/var_expand.md:48]: const also

**Review by sirainen — CHANGES_REQUESTED** [review]

**Review by sirainen — CHANGES_REQUESTED** [review]

File: docs/core/settings/variables.md [hunk 27]
```diff
@@ -27,24 +30,113 @@ You can use special variables in several places:
...
+
+When value is missing or empty, you can use the `default` filter to provide value. Missing variables
+will cause errors and must be negated with default. This does not apply to all providers, some
+providers return empty when value is missing.
+
+If the last filter would output binary data, the data is encoded with `hexlify` filter by default.
+To avoid this, you can use `text` filter, which will sanitize the input and mark it as text.
+
+The new syntax also supports simple maths, you can do one operation. E.g. `%{port + 1000}`.
+Addition, substraction, multiplication, division and modulo operations are supported for now.
```
[the remark this answers is not in the document]
Comment (cmouse) [reply docs/core/settings/variables.md:71]: not supported right now, it only supports *one* operation, as stated in the documentation.

File: docs/core/settings/variables.md [hunk 28]
```diff
@@ -27,24 +30,113 @@ You can use special variables in several places:
...
+## List of filters
+
+All parameters are strings unless stated otherwise.
+If parameter is any, it accepts both numbers and strings.
+Boolean type is `0` for false and `1` for true.
+
+| Filter | Description |
+| ------ | ----------- |
+| base64(pad=boolean, url=boolean) | Base64 encode given input, defaults to pad and not url scheme. |
+| benumber | Convert big-endian encoded number into a number. |
```
[the remark this answers is not in the document]
Comment (cmouse) [reply docs/core/settings/variables.md:85]: idk better word, it means that it will interpret \x04\x00\x00\x00 as 4.

[the remark this answers is not in the document]
Comment (cmouse) [reply docs/core/settings/variables.md:85]: ah, i guess input.

File: docs/core/settings/variables.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (sirainen) [reply docs/core/settings/variables.md:71]: Ah. The "one" word is rather easy to miss or not see its relevance. Should have a separate sentence about it. Like "Only one operation is allowed per expression." Although I wonder what exactly that means. Can I still use some `func(value1 + 1, value2 + 2)`? Or `func(foo + 1) + 2)`?

File: docs/core/config/mailbox/mail_location.md [hunk 29]
```diff
@@ -27,29 +27,15 @@ See [[variable]] for a full list, but the most commonly used ones are:
...
-* `%M` returns a MD5 hash of the string as hex. This can be used for two
-  level hashing by getting substrings of the MD5 hash.
-
-  * Example: `%1Mu/%2.1Mu/%u` returns directories from `0/0/user` to
-    `f/f/user`.
-
-* `%H` returns a 32bit hash of the given string as hex.
-
-  * This is the old, deprecated method. `%N` should be used instead.
+* `%{ user | md5 % 256 | hex}` would return maximum 256 different hashes in range `00` to `ff`.
```
Comment (sirainen) [inline docs/core/config/mailbox/mail_location.md:38]: hex(2) to get `00`

File: docs/core/settings/variables.md [hunk 30]
```diff
@@ -27,24 +30,118 @@ You can use special variables in several places:
...
+A special case for modulo operation is that it can be applied to binary input, e.g. `md5 % 256`.
+The input is treated as 64-bit unsigned number and modulo is taken from that.
+
+All strings must be encapsulated with `"` or `'`, and you can escape them using `\\` within string.
+Numbers when used as parameters must be provided without quotes.
+
+## List of filters
+
+All parameters are strings unless stated otherwise.
+If parameter is `any`, it accepts both numbers and strings.
```
Comment (sirainen) [inline docs/core/settings/variables.md:82]: and bytes?

File: docs/core/settings/variables.md [hunk 31]
```diff
@@ -27,24 +30,118 @@ You can use special variables in several places:
...
+Boolean type is `0` for false and `1` for true.
+
+Filters that have `any` in input and output mean that they will accept bytes or string, and the result
+will be bytes or strings, depending on the input.
```
Comment (sirainen) [inline docs/core/settings/variables.md:86]: What about numbers?

File: docs/core/settings/variables.md [hunk 32]
```diff
@@ -27,24 +30,118 @@ You can use special variables in several places:
...
+| Filter | Input | Output | Description |
+| ------ | ----- | ------ | ----------- |
+| base64(pad=boolean, url=boolean) | String | String | Base64 encode given input, defaults to pad and not url scheme. |
+| benumber | Byte | String | Convert big-endian encoded input into a number. |
+| concat(any, any...) | Byte | Byte | Concatenates input with value(s). Numbers are coerced to strings. Input is optional. |
+| default(value) | String | String | Replace empty or missing input with value. Clears missing variable error. If no value is provided, empty string is used. |
+| domain | String | String | Provides domain part of user@domain value. |
+| hash(method, rounds=number, salt=string) | Byte | Byte | Returns raw hash from input using given hash method. Rounds and salt are optional. |
+| hexlify(width) | Byte | String | Convert bytes into hex with optional width, truncates or pads up to width. |
+| hex(width) | Number | String | Convert base-10 number to base-16 number. If width is specified the result is truncated or padded with 0 to width. Negative width is applied after number. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:97]: This is the input filter that uses Number as input/output. That doesn't seem right. Shouldn't at least benumber, lenumber have Number output?

File: docs/core/settings/variables.md [hunk 33]
```diff
@@ -27,24 +30,118 @@ You can use special variables in several places:
...
+Boolean type is `0` for false and `1` for true.
+
+Filters that have `any` in input and output mean that they will accept bytes or string, and the result
+will be bytes or strings, depending on the input.
+
+| Filter | Input | Output | Description |
+| ------ | ----- | ------ | ----------- |
+| base64(pad=boolean, url=boolean) | String | String | Base64 encode given input, defaults to pad and not url scheme. |
+| benumber | Byte | String | Convert big-endian encoded input into a number. |
+| concat(any, any...) | Byte | Byte | Concatenates input with value(s). Numbers are coerced to strings. Input is optional. |
```
Comment (sirainen) [inline docs/core/settings/variables.md:92]: Byte makes it sound like it's a single 8bit byte. How about Bytes?

**Review by sirainen — CHANGES_REQUESTED** [review]

File: docs/core/settings/variables.md (hunk 30 above)
Comment (cmouse) [reply docs/core/settings/variables.md:82]: there is no special bytes type for variables, variables are either numbers or strings.

File: docs/core/config/optimization.md [hunk 34]
```diff
@@ -71,7 +71,7 @@ the most useful optimization you can do.
 - [[setting,mail_prefetch_count]] setting may be helpful with some mailbox
   formats.
 
-- [[setting,mail_volatile_path,/tmp/dovecot-volatile/%2.256Nu/%u]] moves, e.g.,
+- [[setting,mail_volatile_path,/tmp/dovecot-volatile/%{user | md5 % 256 | hex(4)}/%{user}]] moves, e.g.,
```
Comment (sirainen) [inline docs/core/config/optimization.md:74]: hex(2)

**Review by sirainen — CHANGES_REQUESTED** [review]

File: docs/core/config/mailbox/mail_location.md [hunk 35]
```diff
@@ -27,29 +27,15 @@ See [[variable]] for a full list, but the most commonly used ones are:
...
-| `%n` | User part in `user@domain`; same as `%u` if there's no domain. |
-| `%d` | Domain part in `user@domain`; empty if there's no domain. |
+| `%{user}` | Full username. |
+| `%{user | username}` | User part in `user@domain`; same as `%{user}` if there's no domain. |
+| `%{user | domain}` | Domain part in `user@domain`; empty if there's no domain. |
 
 ### Directory Hashing
 
-You can use three different kinds of hashes in [[variable]].
+Examples on how to do it:
```
Comment (sirainen) [inline docs/core/config/mailbox/mail_location.md:36]: This could also link to the %N upgrading

**Review by sirainen — CHANGES_REQUESTED** [review]

**Review by sirainen — APPROVED** [review]
